File: src/test/kotlin/codes/seanhenry/mockgenerator/generator/templates/GenericReturnValueTest.kt
package codes.seanhenry.mockgenerator.generator.templates

import codes.seanhenry.mockgenerator.entities.Method
import codes.seanhenry.mockgenerator.generator.MockTransformer

class GenericReturnValueTest : MockGeneratorTestTemplate() {

  override fun build(generator: MockTransformer) {
    generator.add(
        Method.Builder("generic1")
            .genericParameter("T")
            .returnType("T")
            .build(),
        Method.Builder("generic2")
            .genericParameter("T")
            .returnType().optional { it.type("T") }
            .build(),
        Method.Builder("generic3")
            .genericParameter("T")
            .returnType().optional { it.unwrapped().type("T") }
            .build(),
        Method.Builder("genericArray")
            .genericParameter("T")
            .returnType().array { it.type("T") }
            .build()
    )
  }
}

File: src/test/resources/stub/GenericReturnValueTest.swift
var stubbedGeneric1Result: Any!
func generic1<T>() -> T {
return stubbedGeneric1Result as! T
}
var stubbedGeneric2Result: Any!
func generic2<T>() -> T? {
return stubbedGeneric2Result as? T
}
var stubbedGeneric3Result: Any!
func generic3<T>() -> T! {
return stubbedGeneric3Result as? T
}
var stubbedGenericArrayResult: [Any]! = []
func genericArray<T>() -> [T] {
return stubbedGenericArrayResult as! [T]
}